controller/controlplane/controller.go

```diff
@@ -489,3 +470,80 @@ func (r *Reconciler) patchStatus(ctx context.Context, logger logr.Logger, update
 
 	return ctrl.Result{}, nil
 }
+
+func (r *Reconciler) ensureWebhookResources(
+	ctx context.Context, logger logr.Logger, cp *operatorv1beta1.ControlPlane,
+) (string, op.Result, error) {
+	webhookEnabled := isAdmissionWebhookEnabled(ctx, r.Client, logger, cp)
+	if !webhookEnabled {
+		log.Debug(logger, "admission webhook disabled, ensuring admission webhook resources are not present", cp)
+	} else {
+		log.Debug(logger, "admission webhook enabled, enforcing admission webhook resources", cp)
+	}
+
+	log.Trace(logger, "ensuring admission webhook service", cp)
+	res, admissionWebhookService, err := r.ensureAdmissionWebhookService(ctx, logger, r.Client, cp)
+	if err != nil {
+		return "", res, fmt.Errorf("failed to ensure admission webhook service: %w", err)
+	}
+	if res != op.Noop {
+		if !webhookEnabled {
+			log.Debug(logger, "admission webhook service has been removed", cp)
+		} else {
+			log.Debug(logger, "admission webhook service has been created/updated", cp)
+		}
+		return "", res, nil // requeue will be triggered by the creation or update of the owned object
+	}
+
+	log.Trace(logger, "ensuring admission webhook certificate", cp)
+	res, admissionWebhookCertificateSecret, err := r.ensureAdmissionWebhookCertificateSecret(ctx, logger, cp, admissionWebhookService)
+	if err != nil {
+		return "", res, err
+	}
+	if res != op.Noop {
+		if !webhookEnabled {
+			log.Debug(logger, "admission webhook service certificate has been removed", cp)
+		} else {
+			log.Debug(logger, "admission webhook service certificate has been created/updated", cp)
+		}
+		return "", res, nil // requeue will be triggered by the creation or update of the owned object
+	}
+
+	log.Trace(logger, "ensuring admission webhook configuration", cp)
+	res, err = r.ensureValidatingWebhookConfiguration(ctx, cp, admissionWebhookCertificateSecret, admissionWebhookService)
+	if err != nil {
+		return "", res, err
+	}
+	if res != op.Noop {
+		if !webhookEnabled {
+			log.Debug(logger, "ValidatingWebhookConfiguration has been removed", cp)
+		} else {
+			log.Debug(logger, "ValidatingWebhookConfiguration has been created/updated", cp)
+		}
+	}
+	if webhookEnabled {
+		return admissionWebhookCertificateSecret.Name, res, nil
+	}
+	return "", res, nil
+}
+
+func isAdmissionWebhookEnabled(ctx context.Context, cl client.Client, logger logr.Logger, cp *operatorv1beta1.ControlPlane) bool {
```

Comment:
Have you considered another API for that, e.g. adding an explicit toggle `DisableValidatingWebhook` to the CP specification? Given this is a net new functionality (disabling the webhook), having it exposed properly via strongly typed API would make sense to me. Using the Pod template spec patch to disable the webhook by setting an env var name that I need to know from somewhere seems like not the best UX to me.

Reply:
Yes, I have (that is briefly mentioned in #308's Proposed solution section).

Assuming that this field would be enabled by default users would have to disable the webhook both via that field and in the env (unless we provide some sort of admission rules that would reject `ControlPlane`s that have those 2 conflicted)

The problem with that is that we'd allow users to create conflicted configurations:

- disable the dedicated field but forget to disable the env: `ControlPlane` will go into `CrashLoopBackoff` due to missing webhook certs in the container
- do not use the dedicated field (leave it enabled by default) but disable the webhook via env: `ControlPlane` will have its webhook resources created but the webhook won't be functional. This would cause all the admission to fail but [get ignored due to our policy setting](https://github.com/Kong/gateway-operator/blob/3e2ae4b34d78e97a83a9c5128a3f5b66b4008f7c/pkg/utils/kubernetes/resources/validatingwebhookconfig/zz_generated_kic_ge3_1.go#L21-L24).

---

Do you believe we should consider a dedicated field given those challenges?

Reply:
> disable the dedicated field but forget to disable the env: ControlPlane will go into CrashLoopBackoff due to missing webhook certs in the container

I assume that, by default, `ControlPlane` doesn't have to specify any Pod template spec patches - its basic functionality should be achievable without customization. Am I wrong?

> do not use the dedicated field (leave it enabled by default) but disable the webhook via env: ControlPlane will have its webhook resources created but the webhook won't be functional. This would cause all the admission to fail but [get ignored due to our policy setting](https://github.com/Kong/gateway-operator/blob/3e2ae4b34d78e97a83a9c5128a3f5b66b4008f7c/pkg/utils/kubernetes/resources/validatingwebhookconfig/zz_generated_kic_ge3_1.go#L21-L24).

Yeah, that sounds like a valid argument which is also valid for any other functionality we allow modifying with the Pod template spec patches (users without Kong's university degree can shoot their feet easily by manually modifying arbitrary env variables in ways unexpected by us). Having this exposed via a strongly typed API would allow us to add a validation that would prevent setting this specific env var via patches. 

I don't want to halt this, but I believe that is something worth discussing. I'm worried we're going to end up with APIs that are not easy to use. If we can avoid relying on patches that are quite unintuitive to me (IMO they should only be used by "power users" for use cases that we haven't thought of yet, or haven't implemented yet), I think we should at least consider that.

Reply:
> I assume that, by default, `ControlPlane` doesn't have to specify any Pod template spec patches - its basic functionality should be achievable without customization. Am I wrong?

That scenario could happen when someone disables the webhook as part of an update.

> I don't want to halt this, but I believe that is something worth discussing. I'm worried we're going to end up with APIs that are not easy to use. If we can avoid relying on patches that are quite unintuitive to me (IMO they should only be used by "power users" for use cases that we haven't thought of yet, or haven't implemented yet), I think we should at least consider that.

💯 I believe that `podTemplateSpec` implementation that we provided was done a bit too far reaching from the get go but now we're stuck with it.

That's something I'd consider for v2 API (this is probably not happening for v1): namely, to get rid of the `podTemplateSpec` and provide a limited number for dedicated fields which we have more control over.

Reply:
I think we should be very careful about the addition of such high-level knobs. ControlPlane and DataPlane are low-level APIs and adding high-level abstractions to low-level settings in the same API (or in different APIs) is a bad idea in my opinion, as it opens the doors to many potential problems, such as the one already discussed in this thread.

Reply:
I think that my main concern here is that we add a functionality (possibility to forcibly not enable CP validating webhook) via a very implicit API (you need to know what specific env variable needs to be set to `off` in order to trigger this quite non-trivial side-effect). Reading the API specification, users won't be able to find it easily I'm afraid. 

Another side of the coin is that we primarily need this feature for passing conformance tests. The intention may be to actually _not_ expose this toggle to the users openly, thus "hiding" this somewhat within the patches' infinite possibilities 😄 If that's the case, I'm fine with doing it as it is.